-1

**Failed Tests**:   Build 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-00ce9c/39206/summary.html
**COMMIT**: 99410bd31e96ed398b48d85729a447ea9036ed13
**CMSSW**: CMSSW_14_1_X_2024-05-01-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/44885/39206/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling edm plugin src/RecoLocalCalo/CaloTowersCreator/src/CaloTowersReCreator.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DDD4HEP_USE_GEANT4_UNITS=1 -DCMSSW_GIT_HASH='CMSSW_14_1_X_2024-05-01-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_14_1_X_2024-05-01-2300' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/cms/cmssw/CMSSW_14_1_X_2024-05-01-2300/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/dd4hep/v01-27-02-4ef356be084e49209272ac38a550fe60/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/alpaka/1.1.0-7d0324257db47fde2d27987e7ff98fb4/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/pcre/8.43-37eb2e8b73bab83d6645ecfd5d73dcaa/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/boost/1.80.0-60a217837b5db1cff00c7d88ec42f53a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/clhep/2.4.7.1-b83201517313ab99b5a082263252c4b6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/cuda/12.4.1-06cde0cd9f95a73a1ea05c8535f60bde/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/gsl/2.6-dc093739c7d4283cbe37712ffdbef83e/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/lcg/root/6.30.07-af9aa5891b3bdefde46d3b3a76481356/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/tbb/v2021.9.0-5849be8e21b090e14f1b189539cee138/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/cms/vdt/0.4.3-fdec6527debaef0832d14d6d478af6ce/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/zlib/1.2.11-51072030b7f93c3ac6c4235f21e413cb/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-ff59ad69bf3dc401976591897b386b6a/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/fmt/8.0.1-786658da8424a36fd9daec568588b7aa/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/OpenBLAS/0.3.15-15504a5c800a9ecbbc2befbb991bead5/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/tinyxml2/6.2.0-b248972f299359d73b1605c96291455f/include -O2 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++17 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -msse3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -Wno-error=unused-variable -DALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128 -DALPAKA_DISABLE_VENDOR_RNG -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/RecoLocalCalo/CaloTowersCreator/src/RecoLocalCaloCaloTowersCreator/CaloTowersReCreator.cc.d src/RecoLocalCalo/CaloTowersCreator/src/CaloTowersReCreator.cc -o tmp/el8_amd64_gcc12/src/RecoLocalCalo/CaloTowersCreator/src/RecoLocalCaloCaloTowersCreator/CaloTowersReCreator.cc.o
>> Compiling edm plugin src/RecoLocalCalo/CaloTowersCreator/src/HcalMaterials.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DDD4HEP_USE_GEANT4_UNITS=1 -DCMSSW_GIT_HASH='CMSSW_14_1_X_2024-05-01-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_14_1_X_2024-05-01-2300' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/cms/cmssw/CMSSW_14_1_X_2024-05-01-2300/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/dd4hep/v01-27-02-4ef356be084e49209272ac38a550fe60/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/alpaka/1.1.0-7d0324257db47fde2d27987e7ff98fb4/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/pcre/8.43-37eb2e8b73bab83d6645ecfd5d73dcaa/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/boost/1.80.0-60a217837b5db1cff00c7d88ec42f53a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/clhep/2.4.7.1-b83201517313ab99b5a082263252c4b6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/cuda/12.4.1-06cde0cd9f95a73a1ea05c8535f60bde/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/gsl/2.6-dc093739c7d4283cbe37712ffdbef83e/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/lcg/root/6.30.07-af9aa5891b3bdefde46d3b3a76481356/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/tbb/v2021.9.0-5849be8e21b090e14f1b189539cee138/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/cms/vdt/0.4.3-fdec6527debaef0832d14d6d478af6ce/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/zlib/1.2.11-51072030b7f93c3ac6c4235f21e413cb/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-ff59ad69bf3dc401976591897b386b6a/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/fmt/8.0.1-786658da8424a36fd9daec568588b7aa/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/OpenBLAS/0.3.15-15504a5c800a9ecbbc2befbb991bead5/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02835/el8_amd64_gcc12/external/tinyxml2/6.2.0-b248972f299359d73b1605c96291455f/include -O2 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++17 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -msse3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -Wno-error=unused-variable -DALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128 -DALPAKA_DISABLE_VENDOR_RNG -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/RecoLocalCalo/CaloTowersCreator/src/RecoLocalCaloCaloTowersCreator/HcalMaterials.cc.d src/RecoLocalCalo/CaloTowersCreator/src/HcalMaterials.cc -o tmp/el8_amd64_gcc12/src/RecoLocalCalo/CaloTowersCreator/src/RecoLocalCaloCaloTowersCreator/HcalMaterials.cc.o
src/RecoLocalCalo/CaloTowersCreator/src/CaloTowersCreationAlgo.cc: In member function 'void CaloTowersCreationAlgo::convert(const CaloTowerDetId&, const MetaTower&, CaloTowerCollection&)':
src/RecoLocalCalo/CaloTowersCreator/src/CaloTowersCreationAlgo.cc:1012:16: error: lvalue required as left operand of assignment
 1012 |       towerP4[3] = 1.f;  // energy
      |       ~~~~~~~~~^
src/RecoLocalCalo/CaloTowersCreator/src/CaloTowersCreationAlgo.cc:1029:17: error: lvalue required as left operand of assignment
 1029 |           emP4[3] = 1.f;  // energy
      |           ~~~~~~^
</pre>

--- a/RecoLocalCalo/CaloTowersCreator/src/CaloTowersCreationAlgo.cc
+++ b/RecoLocalCalo/CaloTowersCreator/src/CaloTowersCreationAlgo.cc
@@ -1009,7 +1009,7 @@ void CaloTowersCreationAlgo::convert(const CaloTowerDetId& id, const MetaTower& 
     case 0: {  // Simple 4-momentum assignment
       GlobalPoint p = theTowerGeometry->getGeometry(id)->getPosition();
       towerP4 = p.basicVector().unit();
-      towerP4[3] = 1.f;  // energy
+      towerP4.mathVector()[3] = 1.f;  // energy
       towerP4 *= E;
 
       // double pf=1.0/cosh(p.eta());
@@ -1026,7 +1026,7 @@ void CaloTowersCreationAlgo::convert(const CaloTowerDetId& id, const MetaTower& 
         if (E_em > 0) {
           emPoint = emShwrPos(metaContains, momEmDepth, E_em);
           emP4 = emPoint.basicVector().unit();
-          emP4[3] = 1.f;  // energy
+          emP4.mathVector()[3] = 1.f;  // energy
           towerP4 = emP4 * E_em;
 
           // double emPf = 1.0/cosh(emPoint.eta());
@@ -1036,7 +1036,7 @@ void CaloTowersCreationAlgo::convert(const CaloTowerDetId& id, const MetaTower& 
           massless = (E_em <= 0);
           hadPoint = hadShwrPos(id, momHadDepth);
           auto lP4 = hadPoint.basicVector().unit();
-          lP4[3] = 1.f;  // energy
+          lP4.mathVector()[3] = 1.f;  // energy
           if (!massless) {
             auto diff = lP4 - emP4;
             mass2 = std::sqrt(E_em * E_had_tot * diff.mag2());
@@ -1057,7 +1057,7 @@ void CaloTowersCreationAlgo::convert(const CaloTowerDetId& id, const MetaTower& 
       } else {  // forward detector: use the CaloTower position
         GlobalPoint p = theTowerGeometry->getGeometry(id)->getPosition();
         towerP4 = p.basicVector().unit();
-        towerP4[3] = 1.f;  // energy
+        towerP4.mathVector()[3] = 1.f;  // energy
         towerP4 *= E;
         // double pf=1.0/cosh(p.eta());
         // if (E>0) towerP4 = CaloTower::PolarLorentzVector(E*pf, p.eta(), p.phi(), 0);  // simple momentum assignment, same position
@@ -1074,7 +1074,7 @@ void CaloTowersCreationAlgo::convert(const CaloTowerDetId& id, const MetaTower& 
         else
           emPoint = theTowerGeometry->getGeometry(id)->getPosition();
         towerP4 = emPoint.basicVector().unit();
-        towerP4[3] = 1.f;  // energy
+        towerP4.mathVector()[3] = 1.f;  // energy
         towerP4 *= E;
 
         // double sumPf = 1.0/cosh(emPoint.eta());
@@ -1084,7 +1084,7 @@ void CaloTowersCreationAlgo::convert(const CaloTowerDetId& id, const MetaTower& 
       } else {  // forward detector: use the CaloTower position
         GlobalPoint p = theTowerGeometry->getGeometry(id)->getPosition();
         towerP4 = p.basicVector().unit();
-        towerP4[3] = 1.f;  // energy
+        towerP4.mathVector()[3] = 1.f;  // energy
         towerP4 *= E;
 
         // double pf=1.0/cosh(p.eta());

